drivers/ms5611/MS5611.cpp

```diff
@@ -403,55 +478,149 @@ int MS5611::_collect(uint32_t &raw)
 
 void MS5611::_measure()
 {
-	if (m_measure_phase == 0) {
-		if (_collect(m_temperature_from_sensor) < 0) {
-			DF_LOG_ERR("error: temp collect failed");
-			reset();
-
-			/* collect fails, re-initiate a temperature read command
-			 * or we are stuck.
-			 */
-			if (_request(ADDR_CMD_CONVERT_D2) < 0) {
-				DF_LOG_ERR("error: temp measure failed");
-			}
+	struct timespec next_wakeup;
+	int ret;
 
-			return;
-		}
+	uint32_t temperature_from_sensor;
+	uint32_t pressure_from_sensor;
 
-		// Request to convert the pressure
-		if (_request(ADDR_CMD_CONVERT_D1) < 0) {
-			DF_LOG_ERR("error: pressure measure failed");
-		}
+	if (_request(ADDR_CMD_CONVERT_D2) < 0) {
+		DF_LOG_ERR("error: temp measure failed");
+		return;
+	}
 
-		m_measure_phase++;
+	clock_gettime(CLOCK_MONOTONIC, &next_wakeup);
+	timespec_inc(&next_wakeup, MS5611_CONVERT_TIME_US * 1000);
+	ret = clock_nanosleep(CLOCK_MONOTONIC, TIMER_ABSTIME, &next_wakeup, NULL);
+	if (ret == EINTR) {
+		DF_LOG_ERR("MS5611 (%d) someone sent me a signal", __LINE__);
+		return;
+	} else if (ret != 0) {
+		DF_LOG_ERR("MS5611 (%d) failed to sleep: %d", __LINE__, ret);
+		return;
+	}
 
-	} else {
-		if (_collect(m_pressure_from_sensor) < 0) {
-			DF_LOG_ERR("error: pressure collect failed");
-			reset();
+	if (_collect(temperature_from_sensor) < 0) {
+		DF_LOG_ERR("error: temp collect failed");
+		reset();
+		return;
+	}
 
-			/* collect fails, re-initiate a pressure read command
-			 * or we are stuck.
-			 */
-			if (_request(ADDR_CMD_CONVERT_D1) < 0) {
-				DF_LOG_ERR("error: pressure measure failed");
-			}
+	if (temperature_from_sensor == 0) {
+		DF_LOG_ERR("MS5611: invalid temperature sample!");
+		return;
+	}
 
-			return;
-		}
+	// Request to convert the pressure
+	if (_request(ADDR_CMD_CONVERT_D1) < 0) {
+		DF_LOG_ERR("error: pressure measure failed");
+	}
+
+	clock_gettime(CLOCK_MONOTONIC, &next_wakeup);
+	timespec_inc(&next_wakeup, MS5611_CONVERT_TIME_US * 1000);
+	ret = clock_nanosleep(CLOCK_MONOTONIC, TIMER_ABSTIME, &next_wakeup, NULL);
+	if (ret == EINTR) {
+		DF_LOG_ERR("MS5611 (%d) someone sent me a signal", __LINE__);
+		return;
+	} else if (ret != 0) {
+		DF_LOG_ERR("MS5611 (%d) failed to sleep: %d", __LINE__, ret);
+		return;
+	}
+
+	if (_collect(pressure_from_sensor) < 0) {
+		DF_LOG_ERR("error: pressure collect failed");
+		reset();
+
+		return;
+	}
+
+	if (pressure_from_sensor == 0) {
```

Comment:
Collect only does a read transaction on the bus - I would rather keep the logic inside _measure.
We can change this, but I wanted to keep the changes as small as possible